### PR TITLE
Enforce coverage

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add tests to enforce the coverage of new python code in the pyccel/ folder.

When test are passing the following summary is shown:

image

When coverage is missing the following summary is shown:

image

**Commit Summary**
- Update `actions/setup-python` to avoid coming deprecation problems
- Update `actions/checkout` to avoid coming deprecation problems
- Add tools for examining git diff and coverage output
- Add job to enforce coverage of new code

### CODE REVIEW

**Comment by EmilyBourne** [comment]
Test

https://github.com/pyccel/pyccel/blob/7ea2526c76a7d3a2d4e63c8d4c86c3c993c8623f/.github/workflows/Github_pytest.yml#L4-L8

File: .github/workflows/Github_pytest.yml [hunk 1]
```diff
@@ -108,3 +114,35 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Python 3.7
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.7
+      - name: Install python dependencies
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install defusedxml
```
Comment (yguclu) [inline .github/workflows/Github_pytest.yml:132]: Good choice!

File: ci_tools/compare_coverage_to_diff.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    ----------
+    coverage_filename : str
+        The name of the xml file containing the coverage information
+
+    Result
+    ------
+    coverage : dict
+            A dictionary whose keys are the files in pyccel
+            and whose values are lists containing the line numbers
+            where coverage is lacking in that file
```
Comment (yguclu) [inline ci_tools/compare_coverage_to_diff.py:22]: This function returns two dictionaries: `changes` and `content_lines`.

File: ci_tools/compare_coverage_to_diff.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    Parse a coverage xml file and return a dictionary containing the files and lines
+    which are untested
+
+    Parameters
+    ----------
+    coverage_filename : str
+        The name of the xml file containing the coverage information
+
+    Result
+    ------
```
Comment (yguclu) [inline ci_tools/compare_coverage_to_diff.py:18]: I think the correct section title is `Returns`

File: ci_tools/compare_coverage_to_diff.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            where coverage is lacking in that file
+    diff     : dict
+            A dictionary whose keys are files which have been
+            changed in this branch and whose values are a dictionary.
+            The dictionary must contain a key 'addition' whose value
+            is a list containing the line numbers of lines which have
+            been changed/added
+
+    Result
+    ------
```
Comment (yguclu) [inline ci_tools/compare_coverage_to_diff.py:60]: I think the correct section title is `Returns`

File: ci_tools/compare_coverage_to_diff.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    equivalent dictionary without lines designed to raise errors
+
+    Parameter
+    ---------
+    untested : dict
+            A dictionary whose keys are the files in pyccel with
+            untested lines which have been added in this branch
+            and whose values are lists containing the line numbers
+            where coverage is lacking in that file
+    """
```
Comment (yguclu) [inline ci_tools/compare_coverage_to_diff.py:89]: Missing `Returns` section

File: ci_tools/git_evaluation_tools.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    """
+    A function which converts the output of a reduced git diff call
+    to a dictionary that can be exported using json.
+    The diff call should use the argument `--unified=0`
+
+    Parameters
+    ----------
+    filename : str
+            The file where the diff was printed
+    """
```
Comment (yguclu) [inline ci_tools/git_evaluation_tools.py:14]: Missing `Returns` section

File: ci_tools/compare_coverage_to_diff.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
+""" Functions for comparing coverage output and git diff output
+"""
```
Comment (yguclu) [inline ci_tools/compare_coverage_to_diff.py:2]: I suggest to change the name of this file (e.g. to `coverage_analysis_tools.py`), in order to make it clear that it is just a module (like `git_evaluation_tools.py`) and not a script (like `check_new_coverage.py`).

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent job! I have a few minor comments

File: .github/workflows/Github_pytest.yml (hunk 1 above)
Comment (EmilyBourne) [reply .github/workflows/Github_pytest.yml:132]: Pylint insists on it

File: ci_tools/compare_coverage_to_diff.py (hunk 7 above)
Comment (EmilyBourne) [reply ci_tools/compare_coverage_to_diff.py]: Fixed 792989a1

File: ci_tools/git_evaluation_tools.py (hunk 6 above)
Comment (EmilyBourne) [reply ci_tools/git_evaluation_tools.py]: Fixed 44267e7a

File: ci_tools/compare_coverage_to_diff.py (hunk 5 above)
Comment (EmilyBourne) [reply ci_tools/compare_coverage_to_diff.py]: Fixed 44267e7a

File: ci_tools/compare_coverage_to_diff.py (hunk 4 above)
Comment (EmilyBourne) [reply ci_tools/compare_coverage_to_diff.py]: Fixed 44267e7a

File: ci_tools/compare_coverage_to_diff.py (hunk 3 above)
Comment (EmilyBourne) [reply ci_tools/compare_coverage_to_diff.py]: Fixed 44267e7a

File: ci_tools/compare_coverage_to_diff.py (hunk 2 above)
Comment (EmilyBourne) [reply ci_tools/compare_coverage_to_diff.py]: Fixed 44267e7a

**Review by yguclu — APPROVED** [review]

**Pull request merged**